associated-token-account/program/src/processor.rs

```diff
@@ -56,6 +70,21 @@ pub fn process_create_associated_token_account(
     let spl_token_program_info = next_account_info(account_info_iter)?;
     let spl_token_program_id = spl_token_program_info.key;
 
+    if create_mode == CreateMode::IfNonExistent
+        && associated_token_account_info.owner == spl_token_program_id
+    {
+        let ata_data = associated_token_account_info.data.borrow();
+        if let Ok(associated_token_account) = StateWithExtensions::<Account>::unpack(&ata_data) {
+            if associated_token_account.base.owner != *wallet_account_info.key {
+                return Err(ProgramError::IllegalOwner);
```

Comment:
This error is meant for an unexpected `AccountInfo::owner`.  I'm not sure it's a good idea to overload it here with program-specific meaning.